Implement 'gamepad' Permissions Policy

marcoscaceres opened the pull request:
#### 82e7a7f647642e9e5158ec4bdc8f628f2e86e1e4
<pre>
Implement &apos;gamepad&apos; Permissions Policy
<a href="https://bugs.webkit.org/show_bug.cgi?id=230136">https://bugs.webkit.org/show_bug.cgi?id=230136</a>
<a href="https://rdar.apple.com/83219098">rdar://83219098</a>

Reviewed by Youenn Fablet.

Implements &quot;gampad&quot; permission policy defaulting to &apos;*&apos; for the allow list.
This matches Chrome&apos;s behavior and is also what&apos;s intended to go into the spec:
<a href="https://github.com/w3c/gamepad/pull/156">https://github.com/w3c/gamepad/pull/156</a>

* LayoutTests/http/tests/gamepad/gamepad-allow-attribute.https-expected.txt: Added.
* LayoutTests/http/tests/gamepad/gamepad-allow-attribute.https.html: Added.
* LayoutTests/http/tests/gamepad/resources/gamepad-postmessage.html: Added.
* Source/WebCore/Modules/gamepad/NavigatorGamepad.cpp:
(WebCore::NavigatorGamepad::getGamepads): Deleted.
* Source/WebCore/Modules/gamepad/NavigatorGamepad.h:
* Source/WebCore/html/FeaturePolicy.cpp:
(WebCore::policyTypeName):
(WebCore::processOriginItem):
(WebCore::FeaturePolicy::parse):
(WebCore::FeaturePolicy::allows const):
* Source/WebCore/html/FeaturePolicy.h:

Canonical link: <a href="https://commits.webkit.org/272199@main">https://commits.webkit.org/272199@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b58a693cd0410d7f4c8847617dd51bc354e2e9ac

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30916 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9590 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32593 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33421 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27928 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/31669 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/11931 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/6847 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/27796 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31238 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8076 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27659 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6927 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7092 "Unexpected infrastructure issue, retrying build") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27538 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/34759 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28151 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28012 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33234 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7129 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5204 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31068 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8836 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/27322 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7297 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7838 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7679 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on LayoutTests/http/tests/gamepad/gamepad-allow-attribute.https.html:
```diff
@@ -0,0 +1,131 @@
+        });
+      }
+
+      // Uses '*' (all) as the permissions policy
```
This comment is probably for the default value, so maybe move it to the first test where default is being used.

youennf commented on Source/WebCore/Modules/gamepad/NavigatorGamepad.cpp:
```diff
@@ -91,6 +91,9 @@ const Vector<RefPtr<Gamepad>>& NavigatorGamepad::getGamepads(Navigator& navigato
     }
 
+    if (!isFeaturePolicyAllowedByDocumentAndAllOwners(FeaturePolicy::Type::Gamepad, *document, LogFeaturePolicyFailure::Yes))
+        return Exception { ExceptionCode::SecurityError, "Third-party iframes are not allowed to call getGamepads() unless explicitly allowed via Feature-Policy (gamepad)"_s };
```
This follows the spec so is ok as is.
The specs using permission policy are not throwing/rejecting with the same error (sometimes SecurityError, sometimes NotAllowedError for instance).

marcoscaceres replied:
Yeah, I did file an issue about this a few years back... unfortunately, it hasn't been addressed yet: https://github.com/w3c/webappsec-permissions-policy/issues/396

marcoscaceres committed:
Implement 'gamepad' Permissions Policy
https://bugs.webkit.org/show_bug.cgi?id=230136
rdar://83219098

Reviewed by Youenn Fablet.

Implements "gampad" permission policy defaulting to '*' for the allow list.
This matches Chrome's behavior and is also what's intended to go into the spec:
https://github.com/w3c/gamepad/pull/156

* LayoutTests/http/tests/gamepad/gamepad-allow-attribute.https-expected.txt: Added.
* LayoutTests/http/tests/gamepad/gamepad-allow-attribute.https.html: Added.
* LayoutTests/http/tests/gamepad/resources/gamepad-postmessage.html: Added.
* Source/WebCore/Modules/gamepad/NavigatorGamepad.cpp:
(WebCore::NavigatorGamepad::getGamepads): Deleted.
* Source/WebCore/Modules/gamepad/NavigatorGamepad.h:
* Source/WebCore/html/FeaturePolicy.cpp:
(WebCore::policyTypeName):
(WebCore::processOriginItem):
(WebCore::FeaturePolicy::parse):
(WebCore::FeaturePolicy::allows const):
* Source/WebCore/html/FeaturePolicy.h:

Canonical link: https://commits.webkit.org/272199@main
